Pointer TextMarshaler keys have the MarshalText method called even on nil values

**Describe the bug**

When marshaling maps with keys of pointer kind implementing TextMarshaler, the MarshalText method is always called even if the key value is nil.

**To Reproduce**

Run the following code:
```go
type ITM struct{ s int }

func (i ITM) MarshalText() ([]byte, error) {
	return []byte("u" + strconv.Itoa(i.s)), nil
}

func main() {
	b, err := sonic.ConfigStd.Marshal(map[*ITM]int{nil: 1})
	fmt.Println(string(b), err)
}
```

Then you will get the application panicking for trying to call MarshalText on nil value.
```
panic: value method main.ITM.MarshalText called using nil *ITM pointer

goroutine 1 [running]:
main.(*ITM).MarshalText(0x93d990?)
        <autogenerated>:1 +0x54
github.com/bytedance/sonic/internal/encoder/alg.(*MapIterator).appendConcrete(0x469445?, 0xc00012e400, 0x6976c0?, 0xc000111b58?)
        /home/bimoadityar/.go/pkg/mod/github.com/bytedance/sonic@v1.13.3/internal/encoder/alg/mapiter.go:124 +0x90
github.com/bytedance/sonic/internal/encoder/alg.(*MapIterator).appendGeneric(0xc000111c70?, 0xc00012e400, 0xc00043d380?, 0x69f5a0?, 0x0?)
        /home/bimoadityar/.go/pkg/mod/github.com/bytedance/sonic@v1.13.3/internal/encoder/alg/mapiter.go:113 +0x59a
github.com/bytedance/sonic/internal/encoder/alg.(*MapIterator).append(0x68ce20?, 0x0?, 0xc00012e3c0?, 0xc000111d18?)
        /home/bimoadityar/.go/pkg/mod/github.com/bytedance/sonic@v1.13.3/internal/encoder/alg/mapiter.go:90 +0x8f
github.com/bytedance/sonic/internal/encoder/alg.IteratorStart(0x68ce20, 0xc0003266c0, 0x1)
        /home/bimoadityar/.go/pkg/mod/github.com/bytedance/sonic@v1.13.3/internal/encoder/alg/mapiter.go:187 +0x105
encode_map[...]int()
        ?:0 +0x15f
github.com/bytedance/sonic/internal/encoder/x86.EncodeTypedPointer(0xc000292d38, 0x68ce20, 0xc00012c968, 0xc000400000, 0x1)
        /home/bimoadityar/.go/pkg/mod/github.com/bytedance/sonic@v1.13.3/internal/encoder/x86/stbus.go:48 +0x125
github.com/bytedance/sonic/internal/encoder.encodeInto(0xc000292d38, {0x68ce20, 0xc0003266c0}, 0x1)
        /home/bimoadityar/.go/pkg/mod/github.com/bytedance/sonic@v1.13.3/internal/encoder/encoder.go:221 +0xd7
github.com/bytedance/sonic/internal/encoder.encodeIntoCheckRace(...)
        /home/bimoadityar/.go/pkg/mod/github.com/bytedance/sonic@v1.13.3/internal/encoder/encode_norace.go:23
github.com/bytedance/sonic/internal/encoder.Encode({0x68ce20, 0xc0003266c0}, 0x1)
        /home/bimoadityar/.go/pkg/mod/github.com/bytedance/sonic@v1.13.3/internal/encoder/encoder.go:174 +0x45
github.com/bytedance/sonic.frozenConfig.Marshal(...)
        /home/bimoadityar/.go/pkg/mod/github.com/bytedance/sonic@v1.13.3/sonic.go:101
main.main()
        /home/bimoadityar/Documents/bimo-test-2/cmd/data/main.go:19 +0x73
```

**Expected behavior**

The MarshalText method is not called for nil value, with output `{"":1} <nil>`.

**Sonic version:**

v1.13.3

**Environment:**

```
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/bimoadityar/.cache/go-build"
GOENV="/home/bimoadityar/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/bimoadityar/.go/pkg/mod"
GOOS="linux"
GOPATH="/home/bimoadityar/.go"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/bimoadityar/.config/spkit/bin/sdk/go1.17.13"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/bimoadityar/.config/spkit/bin/sdk/go1.17.13/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.17.13"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/bimoadityar/Documents/bimo-test-2/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1204808054=/tmp/go-build -gno-record-gcc-switches"
```

--- a/internal/encoder/alg/mapiter.go
+++ b/internal/encoder/alg/mapiter.go
@@ -84,15 +84,24 @@ func (self *MapIterator) data() (p []_MapPair) {
 func (self *MapIterator) append(t *rt.GoType, k unsafe.Pointer, v unsafe.Pointer) (err error) {
     p := self.add()
     p.v = v
+    tk := t.Kind()
 
-    /* check for strings */
-    if tk := t.Kind(); tk != reflect.String {
-        return self.appendGeneric(p, t, tk, k)
+    // followed as `encoding/json/emcode.go:resolveKeyName
+    if tk == reflect.String {
+        p.k = *(*string)(k)
+        return nil
     }
 
-    /* fast path for strings */
-    p.k = *(*string)(k)
-    return nil
+    // check if the key implements the encoding.TextMarshaler interface
+    if t.Pack().Implements(vars.EncodingTextMarshalerType) {
+        if tk != reflect.Interface {
+            return self.appendConcrete(p, t, k)
+        } else {
+            return self.appendInterface(p, t, k)
+        }
+    }
+
+    return self.appendGeneric(p, t, tk, k)
 }
 
 func (self *MapIterator) appendGeneric(p *_MapPair, t *rt.GoType, v reflect.Kind, k unsafe.Pointer) error {
@@ -109,34 +118,43 @@ func (self *MapIterator) appendGeneric(p *_MapPair, t *rt.GoType, v reflect.Kind
         case reflect.Uint64    : p.k = rt.Mem2Str(strconv.AppendUint(p.m[:0], uint64(*(*uint64)(k)), 10))  ; return nil
         case reflect.Uintptr   : p.k = rt.Mem2Str(strconv.AppendUint(p.m[:0], uint64(*(*uintptr)(k)), 10)) ; return nil
         case reflect.Bool      : if *(*bool)(k) { p.k = "true" } else { p.k = "false" }; return nil
-        case reflect.Interface : return self.appendInterface(p, t, k)
-        case reflect.Struct, reflect.Ptr : return self.appendConcrete(p, t, k)
         default                : return vars.Error_type(t.Pack())
     }
 }
 
-func (self *MapIterator) appendConcrete(p *_MapPair, t *rt.GoType, k unsafe.Pointer) (err error) {
+func (self *MapIterator) appendConcrete(p *_MapPair, t *rt.GoType, k unsafe.Pointer) error {
     // compiler has already checked that the type implements the encoding.MarshalText interface
     if !t.Indirect() {
         k = *(*unsafe.Pointer)(k)
     }
+
+    // check the TextMarshaler interface
     eface := rt.GoEface{Value: k, Type: t}.Pack()
-    out, err := eface.(encoding.TextMarshaler).MarshalText()
+    e, ok := eface.(encoding.TextMarshaler)
+    if !ok {
+        return vars.Error_type(t.Pack())
+    }
+
+    // check for nil pointer
+    if t.Kind() == reflect.Ptr && k == nil {
+        p.k = ""
+        return nil
+    }
+    
+    out, err := e.MarshalText()
     if err != nil {
         return err
     }
     p.k = rt.Mem2Str(out)
-    return
+    return nil
 }
 
 func (self *MapIterator) appendInterface(p *_MapPair, t *rt.GoType, k unsafe.Pointer) (err error) {
     if len(rt.IfaceType(t).Methods) == 0 {
         panic("unexpected map key type")
-    } else if p.k, err = asText(k); err == nil {
-        return nil
-    } else {
-        return
     }
+    p.k, err = asText(k)
+    return
 }
 
 func IteratorStop(p *MapIterator) {

--- a/internal/encoder/compiler.go
+++ b/internal/encoder/compiler.go
@@ -287,6 +287,12 @@ func (self *Compiler) compileMapBody(p *ir.Program, sp int, vt reflect.Type) {
 }
 
 func (self *Compiler) compileMapBodyKey(p *ir.Program, vk reflect.Type) {
+	// followed as `encoding/json/emcode.go:resolveKeyName
+	if vk.Kind() == reflect.String {
+		self.compileString(p, vk)
+		return
+	}
+
 	if !vk.Implements(vars.EncodingTextMarshalerType) {
 		self.compileMapBodyTextKey(p, vk)
 	} else {

--- a/issue_test/issue827_test.go
+++ b/issue_test/issue827_test.go
@@ -1,0 +1,170 @@
+/**
+ * Copyright 2025 ByteDance Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package issue_test
+
+import (
+	"encoding/json"
+	"strconv"
+	"testing"
+
+	"github.com/bytedance/sonic"
+	"github.com/stretchr/testify/require"
+)
+
+type ITM int
+type ITM8 int8
+type ITM16 int16
+type ITM32 int32
+type ITM64 int64
+type UTM uint
+type UTM8 uint8
+type UTM16 uint16
+type UTM32 uint32
+type UTM64 uint64
+type FTM32 float32
+type FTM64 float64
+type StrTM string
+type Str2TM string
+type CustomTM chan int
+type STM struct{ ID int }
+type PTM struct{ ID int }
+
+func (i ITM) MarshalText() ([]byte, error) {
+	return []byte("u" + strconv.Itoa(int(i))), nil
+}
+
+func (i ITM8) MarshalText() ([]byte, error) {
+	return []byte("u8" + strconv.FormatInt(int64(i), 10)), nil
+}
+
+func (i ITM16) MarshalText() ([]byte, error) {
+	return []byte("u16" + strconv.FormatInt(int64(i), 10)), nil
+}
+
+func (i ITM32) MarshalText() ([]byte, error) {
+	return []byte("u32" + strconv.FormatInt(int64(i), 10)), nil
+}
+
+func (i ITM64) MarshalText() ([]byte, error) {
+	return []byte("u64" + strconv.FormatInt(int64(i), 10)), nil
+}
+
+func (i UTM) MarshalText() ([]byte, error) {
+	return []byte("u" + strconv.FormatUint(uint64(i), 10)), nil
+}
+
+func (i UTM8) MarshalText() ([]byte, error) {
+	return []byte("u8" + strconv.FormatUint(uint64(i), 10)), nil
+}
+
+func (i UTM16) MarshalText() ([]byte, error) {
+	return []byte("u16" + strconv.FormatUint(uint64(i), 10)), nil
+}
+
+func (i UTM32) MarshalText() ([]byte, error) {
+	return []byte("u32" + strconv.FormatUint(uint64(i), 10)), nil
+}
+
+func (i UTM64) MarshalText() ([]byte, error) {
+	return []byte("u64" + strconv.FormatUint(uint64(i), 10)), nil
+}
+
+func (f FTM32) MarshalText() ([]byte, error) {
+	return []byte("f32" + strconv.FormatFloat(float64(f), 'f', -1, 32)), nil
+}
+
+func (f FTM64) MarshalText() ([]byte, error) {
+	return []byte("f64" + strconv.FormatFloat(float64(f), 'f', -1, 64)), nil
+}
+
+func (s STM) MarshalText() ([]byte, error) {
+	return []byte("struct" + strconv.Itoa(s.ID)), nil
+}
+
+func (p *PTM) MarshalText() ([]byte, error) {
+	return []byte("ptr" + strconv.Itoa(p.ID)), nil
+}
+
+func (c CustomTM) MarshalText() ([]byte, error) {
+	if c == nil {
+		return []byte("customnil"), nil
+	}
+	return []byte("custom"), nil
+}
+
+func (s *StrTM) MarshalText() ([]byte, error) {
+	return []byte("str" + string(*s)), nil
+}
+
+func (s Str2TM) MarshalText() ([]byte, error) {
+	return []byte("str" + string(s)), nil
+}
+
+func TestIssue827_AllTextMarshalerTypes(t *testing.T) {
+	p1 := PTM{ID: 1}
+	p2 := PTM{ID: 2}
+
+	s1 := Str2TM("one")
+	s2 := Str2TM("two")
+
+	testCases := []struct {
+		name string
+		data interface{}
+	}{
+		{"int", map[ITM]int{1: 1, 2: 2}},
+		{"int8", map[ITM8]int{1: 1, 2: 2}},
+		{"int16", map[ITM16]int{1: 1, 2: 2}},
+		{"int32", map[ITM32]int{1: 1, 2: 2}},
+		{"int64", map[ITM64]int{1: 1, 2: 2}},
+		{"uint", map[UTM]int{1: 1, 2: 2}},
+		{"uint8", map[UTM8]int{1: 1, 2: 2}},
+		{"uint16", map[UTM16]int{1: 1, 2: 2}},
+		{"uint32", map[UTM32]int{1: 1, 2: 2}},
+		{"uint64", map[UTM64]int{1: 1, 2: 2}},
+		{"float32", map[FTM32]int{1.5: 1, 2.7: 2}},
+		{"float64", map[FTM64]int{1.5: 1, 2.7: 2}},
+		{"struct", map[STM]int{{ID: 1}: 1, {ID: 2}: 2}},
+		{"pointer", map[*PTM]int{&p1: 1, &p2: 2, nil: 3}},
+		{"string", map[StrTM]int{"one": 1, "two": 2}},
+		{"string2", map[*Str2TM]int{&s1: 1, &s2: 2, nil: 3}},
+		{"string3", map[Str2TM]int{s1: 1, s2: 2}},
+		{"custom", map[CustomTM]int{nil: 1, make(chan int): 2}},
+	}
+	
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			jsonData, err := json.Marshal(tc.data)
+			require.NoError(t, err)
+			expected := string(jsonData)
+			println(tc.name, "expected:", expected)
+			
+			sonicData, err := sonic.Marshal(tc.data)
+			require.NoError(t, err)
+			
+			sonicStdData, err := sonic.ConfigStd.Marshal(tc.data)
+			require.NoError(t, err)
+			require.Equal(t, expected, string(sonicStdData))
+
+			// compare the output of sonic.Marshal
+			var sonicMap, jsonMap map[string]int
+			require.NoError(t, json.Unmarshal(sonicData, &sonicMap))
+			require.NoError(t, json.Unmarshal(jsonData, &jsonMap))
+			require.Equal(t, jsonMap, sonicMap)
+			
+		})
+	}
+}

